### ISSUE TITLE
Error with password containing ampersand

### ISSUE THREAD

**Issue by antode** [barbushin/php-imap#203]
Trying to call $mailbox->searchMailbox('ALL'):

```php
$mailbox = new PhpImap\Mailbox(
    '{imap.gmail.com:993/imap/ssl}INBOX', 
    'some@gmail.com', 
    'some&password&with&ampersands',
     __DIR__
);

$mailsIds = $mailbox->searchMailbox('ALL');
```
Causes:
```
PHP Fatal error:  Uncaught PhpImap\ConnectionException: IMAP method imap_open() 
failed with error: Can not authenticate to IMAP server: [AUTHENTICATIONFAILED] 
Invalid credentials (Failure) in 
/var/www/my_project/vendor/php-imap/php-imap/src/PhpImap/Mailbox.php:828
```
Seems like imap_utf7_encode() at PhpImap\Mailbox::imap() on line 814 converts ampersand wrong way.

```
user@user-pc:~$ php -a
Interactive mode enabled

php > var_dump(imap_utf7_encode('some&password&with&ampersands'));
php shell code:1:
string(32) "some&-password&-with&-ampersands"
```

**Comment by bc-moe** [comment]
HI @antode - glad you posted this.  Been struggling with this issue for the last few hours and just realised there was a bug.  I understand it's a bug to be resolved on the next release - but how did you resolve it in the meantime?

**Comment by antode** [comment]
@bc-moe Hi! Unfortunately i didn't, but as a workaround you can define your own class and override imap method PhpImap\MailBox:imap() like this:

```php
class MyMailbox extends PhpImap\Mailbox
{
    public function imap($methodShortName, $args = [], $prependConnectionAsFirstArg = true, $throwExceptionClass = \PhpImap\Exception::class)
    {
        if(!is_array($args)) {
            $args = [$args];
        }
        foreach($args as &$arg) {
            if(is_string($arg)) {
                $arg = imap_utf7_encode($arg);
            }
        }
        if($prependConnectionAsFirstArg) {
            array_unshift($args, $this->getImapStream());
        }

        // Magic here 
        if ($methodShortName === 'open') { 
            $args[2] = imap_utf7_decode($args[2]);
        }

        imap_errors(); // flush errors
        $result = @call_user_func_array("imap_$methodShortName", $args);

        if(!$result) {
            $errors = imap_errors();
            if($errors) {
                if($throwExceptionClass) {
                    throw new $throwExceptionClass("IMAP method imap_$methodShortName() failed with error: " . implode('. ', $errors));
                }
                else {
                    return false;
                }
            }
        }

        return $result;
    }
}

$mailbox = new MyMailbox(
    '{imap.gmail.com:993/imap/ssl}INBOX', 
    'some@gmail.com', 
    'some&password&with&ampersands',
     __DIR__
);

$mailsIds = $mailbox->searchMailbox('ALL');
```

**Comment by antode** [comment]
@barbushin Should we encode all arguments in utf-7? Can we encode only certain arguments in certain methods except of encode all of them in imap()?

**Comment by barbushin** [comment]
@antode 

> Can we encode only certain arguments in certain methods except of encode all of them in imap()?

Sure, but... do you know what exactly args should be encoded, and why?

**Comment by sambenne** [comment]
We have just ran into this problem, our mail server had a password change that contains an & and upgrading to the latest version now breaks. So we are staying in Version 2. For the time being.

**Comment by ghxter** [comment]
Please view help you ? 
https://github.com/barbushin/php-imap/issues/242

**Comment by Sebi94nbg** [comment]
@barbushin can be closed as well: PR #295.
